src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -795,6 +795,15 @@ trait Namers extends MethodSynthesis {
       tree.symbol = enterClassSymbol(tree)
       tree.symbol setInfo completerOf(tree)
 
+      // ensure completion before patmat (unit is not run for java, so add check directly to phase)
+      if (tree.symbol.isJava) {
+        val sym = tree.symbol
+        currentRun.typerPhase match {
+          case ph: typerFactory.TyperPhase => ph.addCheckAfterTyper(() => sym.initialize)
```

Comment:
Forcing every class symbol from a Java source could uncover / force existing bugs / limitations we have in our Java support and lead to compilation failures. Maybe we can restrict it somehow, ideally to symbols used in matches?

Reply:
I was hoping you would have an idea about that. I mean about the somehow part.

Reply:
Actually, I was hoping you would have pushed an idea about that, as you are wont to do.

Reply:
Hmm. Initializing "symbols used in matches" as I suggested above doesn't help. In `(x: A) match { case _: C => ... }`, we need to initialize all children of `A`, not only those that appear in the match.

Reply:
yes, your first thought made me think just force ALL the symbols. I would have to think about your second thought when awake.

Reply:
For `permits` everything seems to be in place already: https://github.com/lrytz/scala/commit/34d2825621026977231feaafb0adad159b1f10ad. Didn't test if the recursion is needed.

We still need a way to find other classes defined in the same compilation unit.

Reply:
OK I'm almost awake. I thought it's always desirable to force as late as possible, that is, after typer is better than during typer.

I recall thinking that the hiccup is that top-level symbols isn't enough. I'll try to come up with the test cases, if I can remember where to put the semicolons in Java.

Reply:
edit: after checking twitter and fosstodon, I noticed that your innovation was to check only from `typedMatch`.